### PR TITLE
TPD-823, TPD-787 Giving Settings Page

### PR DESCRIPTION
**Pull request by jason-rice**
created a migration to create GivingPages table and GivingPageFunds table, if these tables don't already exist in the database schema. Everything appears to be working correctly.
[touchpointsoftware.atlassian.net/browse/TPD-823](touchpointsoftware.atlassian.net/browse/TPD-823)
[touchpointsoftware.atlassian.net/browse/TPD-787](https://touchpointsoftware.atlassian.net/browse/TPD-787)

### CODE REVIEW

File: UnitTests/SharedTestFixtures/DatabaseTestBase.cs [hunk 1]
```diff
@@ -117,6 +117,87 @@ protected Organization CreateOrganization(string name = null, int? fromId = null
...
+        {
+            db.GivingPages.DeleteOnSubmit(givingPage);
+            db.SubmitChanges();
+        }
+
+        protected GivingPageFund CreateGivingPageFund(int givingPageId, int fundId)
+        {
+            var givingPageFund = new GivingPageFund
+            {
+                GivingPageFundId = RandomNumber(),
```
Comment (brondavies) [inline UnitTests/SharedTestFixtures/DatabaseTestBase.cs:155]: This should be generated when you call `SubmitChanges`

File: UnitTests/SharedTestFixtures/DatabaseTestBase.cs [hunk 2]
```diff
@@ -117,6 +117,87 @@ protected Organization CreateOrganization(string name = null, int? fromId = null
             return newOrg;
         }
 
+        protected GivingPage CreateGivingPage(string pageName = null, int? fundId = null)
```
Comment (brondavies) [inline UnitTests/SharedTestFixtures/DatabaseTestBase.cs:120]: Most of the methods you've added here can be placed in a class under the `SharedTestFixtures` project and used like the `MockContributions` class is used statically.

File: UnitTests/SharedTestFixtures/DatabaseTestBase.cs (hunk 1 above)
Comment (jason-rice) [reply UnitTests/SharedTestFixtures/DatabaseTestBase.cs:155]: You're right, fixed this

File: UnitTests/SharedTestFixtures/DatabaseTestBase.cs (hunk 2 above)
Comment (jason-rice) [reply UnitTests/SharedTestFixtures/DatabaseTestBase.cs]: Ok I split the methods out like requested, the testing code is more organized and the DatabaseTestBase.cs class isn't as heavy this way. great suggestion!

File: UnitTests/CmsDataTests/GivingSettings/GivingPageCRUD.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿using CmsData.View;
+using SharedTestFixtures;
+using System.Linq;
+using Xunit;
+
+namespace CmsDataTests.GivingSettings
+{
+    [Collection(Collections.Database)]
+    public class GivingPageCRUD : DatabaseTestBase
```
Comment (brondavies) [inline UnitTests/CmsDataTests/GivingSettings/GivingPageCRUD.cs:9]: Naming convention of the test class (and file) should match the class being tested e.g. `GivingPageTests`

Comment (jason-rice) [reply UnitTests/CmsDataTests/GivingSettings/GivingPageCRUD.cs]: I changed this according to your suggestion

File: CmsWeb/src/components/GenericComponents/Modal.vue [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      </div>
+    </transition>
+  </div>
+</template>
+
+<script>
+export default {
+  data: function() {
+    return {
+      theHeader: "this is the header",
```
Comment (timdavila) [inline CmsWeb/src/components/GenericComponents/Modal.vue:32]: Is this being used? Looks like placeholder

File: CmsWeb/src/App.vue [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<template>
+  <div id="app">
+    <!--<img alt="Vue logo" src="./assets/logo.png">-->
```
Comment (timdavila) [inline CmsWeb/src/App.vue:3]: Remove any commented out code not in use please

File: CmsWeb/Content/dist/index.html [hunk 6]
```diff
@@ -0,0 +1 @@
+<!DOCTYPE html><html><head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width,initial-scale=1"><title>Vue App</title><link href=/css/app.css rel=preload as=style><link href=/js/app.js rel=preload as=script><link href=/css/app.css rel=stylesheet></head><body><div id=app></div><script src=/js/app.js></script></body></html>
```
Comment (timdavila) [inline CmsWeb/Content/dist/index.html:1]: remove this file if it's not being used

File: CmsWeb/Areas/Giving/Models/GivingPageModel.cs [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+            get => _db;
+            set
+            {
+                _db = value;
+            }
+        }
+
+        public List<GivingPageItem> GetGivingPageList()
+        {
+            var outputList = new List<GivingPageItem>();
```
Comment (timdavila) [inline CmsWeb/Areas/Giving/Models/GivingPageModel.cs:32]: This method could use some refactoring. Separate discrete and repeated functions into their own methods, and give all variables clear names

File: CmsWeb/Areas/Giving/Controllers/GivingPageController.cs [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+            }
+        }
+
+        [HttpGet]
+        public JsonResult GetPageTypes()
+        {
+            var pageTypesList = new List<PageTypesClass>();
+            PageTypesClass pledge = new PageTypesClass
+            {
+                id = 1,
```
Comment (timdavila) [inline CmsWeb/Areas/Giving/Controllers/GivingPageController.cs:69]: You can avoid "magic" numbers with enums or a class with hardcoded page types in CmsData.Codes

File: CmsWeb/.gitignore [hunk 9]
```diff
@@ -0,0 +1,22 @@
+.DS_Store
```
Comment (timdavila) [inline CmsWeb/.gitignore:1]: We have this in the parent directory. I think the only thing new here is /dist which can be added to the old gitignore

File: CmsWeb/src/components/Giving/GivingPageIndex/GivingPageIndex.vue [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+                    <a v-on:click="sort('provincia')" style="cursor:pointer">Action</a>
+                  </th>
+                </tr>
+              </thead>
+              <tbody>
+                <tr v-for="(items, index) in givingPageList" :key="items.GivingPageId">
+                  <td>
+                    <generic-slider
+                      v-bind:sliderValue="items.Enabled"
+                      v-bind:class="['toggle', 'btn', items.Enabled ? 'btn-primary' : 'btn-default', items.Enabled ? 'on' : 'off']"
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/src/components/Giving/GivingPageIndex/GivingPageIndex.vue:90]: these bindings are in the slider component, but in order to get css properties to change in real time, when editing a giving page, these bindings are also needed on this component in location.

File: CmsWeb/src/components/UsersGrid/UsersGrid.vue [hunk 11]
```diff
@@ -0,0 +1,47 @@
+﻿<template>
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/src/components/UsersGrid/UsersGrid.vue:1]: removed this component

File: CmsWeb/src/components/UsersGrid/UserCardGrid.vue [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    import axios from 'axios';
+    export default {
+        data: function () {
+            return {
+                users: [],
+            };
+        },
+        methods: {
+            async fetchUsers() {
+                const { data } = await axios.get('https://jsonplaceholder.typicode.com/users');
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/src/components/UsersGrid/UserCardGrid.vue:27]: removed this component

File: CmsWeb/src/components/Giving/GivingPageIndex/GivingPageIndex.vue [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+                    <a v-on:click="sort('provincia')" style="cursor:pointer">Shell</a>
+                  </th>
+                  <th>
+                    <a v-on:click="sort('indirizzo')" style="cursor:pointer">Page Type</a>
+                  </th>
+                  <th>
+                    <a v-on:click="sort('provincia')" style="cursor:pointer">Default Fund</a>
+                  </th>
+                  <th>
+                    <a v-on:click="sort('provincia')" style="cursor:pointer">Action</a>
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/src/components/Giving/GivingPageIndex/GivingPageIndex.vue:81]: removed these parameters, they were placeholders for actually implementing sorting but didn't have time to get to that

File: CmsWeb/src/components/GenericComponents/Slider.vue [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    mySliderValue() {
+      this.$emit("toggleSlider");
+    }
+  },
+  methods: {}
+};
+</script>
+
+<style scoped>
+/* .switch {
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/src/components/GenericComponents/Slider.vue:41]: removed this

File: CmsWeb/src/components/GenericComponents/ShowModalBtn.vue [hunk 15]
```diff
@@ -0,0 +1,117 @@
+<template>
+  <div>
+    <button class="button" @click="showMyModal = true">Show Modal</button>
+    <transition name="fade" appear>
+      <div class="modal-overlay" v-if="showMyModal" @click="showMyModal = false"></div>
+    </transition>
+    <transition name="slide" appear>
+      <div class="modal" v-if="showMyModal">
+        <h1>Lorem Ipsum</h1>
```
[the remark this answers is not in the document]
Comment (jason-rice) [reply CmsWeb/src/components/GenericComponents/ShowModalBtn.vue:9]: removed this

File: CmsWeb/src/components/GenericComponents/Modal.vue (hunk 4 above)
Comment (jason-rice) [reply CmsWeb/src/components/GenericComponents/Modal.vue:32]: removed this

File: CmsWeb/src/App.vue (hunk 5 above)
Comment (jason-rice) [reply CmsWeb/src/App.vue:3]: removed this

File: CmsWeb/Content/dist/index.html (hunk 6 above)
Comment (jason-rice) [reply CmsWeb/Content/dist/index.html:1]: this file is created every time a vue build is run, not sure how to remove this

File: CmsWeb/Areas/Giving/Controllers/GivingPageController.cs (hunk 8 above)
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Controllers/GivingPageController.cs:69]: great suggestion, I fixed this. created a new Giving folder in CmsData.Classes where I put a new class constructor and class implementation on this list.

File: CmsWeb/.gitignore (hunk 9 above)
Comment (jason-rice) [reply CmsWeb/.gitignore:1]: removed this

File: CmsWeb/Areas/Giving/Models/GivingPageModel.cs (hunk 7 above)
Comment (jason-rice) [reply CmsWeb/Areas/Giving/Models/GivingPageModel.cs:29]: i fixed this

**Comment by brondavies** [comment]
Closing in favor of #931

**Pull request closed without merging**